xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -504,9 +533,69 @@ private <T extends ResourceUpdate> void handleResourceUpdate(
     }
   }
 
-  /**
-   * Tracks a single subscribed resource.
-   */
+  @Override
+  public SettableFuture<Void> reportServerConnections(ServerConnectionCallback callback) {
+    SettableFuture<Void> future = SettableFuture.create();
+    syncContext.execute(() -> {
+      serverCpClientMap.forEach((serverInfo, controlPlaneClient) ->
+          callback.reportServerConnectionGauge(
+              controlPlaneClient.hasWorkingAdsStream() ? 1 : 0,
+              target,
+              serverInfo.target()));
+    });
+    future.set(null);
+    return future;
+  }
+
+  @Override
+  public SettableFuture<Void> reportResourceCounts(ResourceCallback callback) {
+    SettableFuture<Void> future = SettableFuture.create();
+    syncContext.execute(() -> {
+      Map<XdsResourceType<?>, Map<String, Long>> resourceCountsByType =
+          getResourceCountsByType();
+      reportResourceCountsToCallback(callback, resourceCountsByType);
```

Comment:
The point of using a callback is mostly to avoid allocating the map. If you're going to construct the map, you can instead just "return" it as the `future` result. See my other comment in getResourceCountsByType().

Although, I'd actually consider getting rid of this method and using `getSubscribedResourcesMetadataSnapshot()` instead. It looks like it has access to all the data needed here and avoids polluting XdsClientImpl with the very specific form of data you need. In particular, "nacked_but_cached" seems highly specific to this one consumer of the data.

(You can contrast this method with `reportServerConnections()`, which works well as a callback as it just loops over the existing data. Although the data it returns is also small, so it could have been fine to copy it.)

Reply:
`getSubscribedResourcesMetadataSnapshot()` is missing `subscriber.data` which tells us if resource is cached and used in determining `grpc.xds.cache_state` attribute value. 
As discussed, I will try adding `subcriber.data` value as a new field in ResourceMetadata and after that I will move the processing / formatting of values to `XdsClientMetricReporterImpl`.

Reply:
You might just add a `boolean` for whether there is data cached. That way it does stay "metadata."

Reply:
Moved resource count by cache state metric aggregating and reporting to `XdsClientmetricReporterImpl". Added a boolean `isCached` to `ResourceMetadata`. PTAL